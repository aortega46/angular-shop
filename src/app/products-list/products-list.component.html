<ul class="grid grid-cols-2 gap-4">
  @for (
    product of productsService.filteredProducts$ | async;
    track product.id
  ) {
    <product-item [product]="product" (onAddToCart)="addToCart($event)" />
  }
</ul>
