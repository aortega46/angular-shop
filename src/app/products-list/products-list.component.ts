import { Component, OnInit, inject } from '@angular/core'
import { ProductItemComponent } from '../product-item/product-item.component'
import { ProductsService } from '../services/products/products.service'
import { AsyncPipe } from '@angular/common'

@Component({
  selector: 'app-products-list',
  standalone: true,
  imports: [ProductItemComponent, AsyncPipe],
  templateUrl: './products-list.component.html',
})
export class ProductsListComponent implements OnInit {
  productsService = inject(ProductsService)

  ngOnInit(): void {
    this.productsService.filterProducts()
  }
}
