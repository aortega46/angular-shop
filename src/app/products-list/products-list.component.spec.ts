import { ComponentFixture, TestBed } from '@angular/core/testing'

import { ProductsListComponent } from './products-list.component'
import { ProductsService } from '../services/products.service'
import { of } from 'rxjs'

describe('ProductsListComponent', () => {
  let component: ProductsListComponent
  let fixture: ComponentFixture<ProductsListComponent>

  let compiled: HTMLElement
  let productsServiceStub: Partial<ProductsService>

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [ProductsListComponent],
      providers: [{ provide: ProductsService, useValue: productsServiceStub }],
    }).compileComponents()

    fixture = TestBed.createComponent(ProductsListComponent)
    component = fixture.componentInstance

    compiled = fixture.nativeElement

    productsServiceStub = {
      filterProducts: () => {},
      filteredProducts$: of([
        { id: 1, title: 'Product 1', price: 20, thumbnail: 'test1.png' },
        { id: 2, title: 'Product 2', price: 30, thumbnail: 'test2.png' },
        { id: 3, title: 'Product 3', price: 15, thumbnail: 'test3.png' },
      ] as any),
    }
  })

  it('should create', () => {
    expect(component).toBeTruthy()
  })

  it('should match snapshot', () => {
    expect(compiled).toMatchSnapshot()
  })

  it('should call filterProducts OnInit', () => {
    const service = TestBed.inject(ProductsService)
    const spyFilterProducts = jest.spyOn(service, 'filterProducts')

    // here calls ngOnInit
    fixture.detectChanges()

    expect(spyFilterProducts).toHaveBeenCalledTimes(1)
  })

  it('should display filteredProducts$', () => {
    fixture.detectChanges()

    const productElements = compiled.querySelectorAll('product-item')
    expect(productElements.length).toBe(3)
    expect(productElements[0].textContent).toContain('Product 1')
    expect(productElements[0].textContent).toContain('$20')
    expect(
      productElements[0].querySelector('img')?.getAttribute('src')
    ).toContain('test1.png')

    expect(productElements[1].textContent).toContain('Product 2')
  })
})
