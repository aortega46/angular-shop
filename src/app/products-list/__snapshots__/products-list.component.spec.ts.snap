// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsListComponent should match snapshot 1`] = `
<div
  id="root1"
>
  <ul
    class="mb-4 grid grid-cols-1 gap-4 sm:grid-cols-2"
  >
    
  </ul>
</div>
`;
